### ISSUE TITLE
GraphViz Error

### ISSUE THREAD

**Issue by ccorcos** [sverweij/dependency-cruiser#15]
Generating the .dot file works, but I'm having trouble getting `dot` to work...

```
 ❯❯❯ dependency-cruise -T dot -x node_modules src  > src.dot
 ❯❯❯ dot -T svg src.dot > src.svg
Assertion failed: (constraining_flat_edge(g,v,e) == FALSE), function flat_reorder, file mincross.c, line 1561.
[1]    6084 abort      dot -T svg src.dot > src.svg
```

Any ideas? Here's my dot file 👍 

[src.dot.zip](https://github.com/sverweij/dependency-cruiser/files/1409847/src.dot.zip)

**Comment by ccorcos** [comment]
Also, super cool project! I see you're a big fan of text-based tools.

**Comment by sverweij** [comment]
Bumping into the limits of graphviz apparently ...
I'll look into it tonight.

=> What version of graphviz are you on?

**Comment by ccorcos** [comment]
Fresh install from homebrew:

```
❯❯❯ dot -V
dot - graphviz version 2.40.1 (20161225.0304)
```

Thanks!

**Comment by sverweij** [comment]
TL;DR the dot dependency-cruiser generates is valid, but the graph is too big for graphviz dot to process (see 'research' below). We might need a different approach.

I did get considerable parts of the graph rendered. The result didn't seem very usable, though - representing >1000 modules and >6000 relations between them probably need a different approach. 
Incidentally #16 might be one of them, but there's probably others as well - the effectiveness depends on how the result is used.

@ccorcos what would you be using the visualisation for?

#### Research
##### Is the generated dot valid? => yep.
- The more unusual thing in the graphs dependency-cruiser produces are the way it declares clusters (so the modules show up in the way they're nested in folders). 
  - So I first removed those leaving only the edges (>6000 of them). This doesn't produce the assertion error, but it takes _dot_  a lot of time. I stopped the process after it ran for ~15min at 100% CPU (core i7). 
  - Secondly I kept the clusters but commented the relations: that renders an (unconnected but nicely nested) graph in no time => _the clusters are probably not the problem_.
- Next I started adding relations 1000 at a time. dot rendered those correctly within reasonable time frames up until there's ~5000 of them. Above dot rapidly gets slower (see below). 

###### dot rendering performance as a function of the number of relations
<img width="671" alt="screen shot 2017-10-25 at 19 53 29" src="https://user-images.githubusercontent.com/4822597/32014450-6ad12c4a-b9be-11e7-89db-47fd9c3e6345.png">

**Comment by ccorcos** [comment]
Interesting. Thanks for the help! My is to try to build an interface for exploring a graph (such as a codebase without using directory structure). The max depth option sounds like a good idea... Perhaps even a JS API to choose certain expansions...

I built something last night using the JSON output -- you might find it interesting (just wrote a readme for you if you want to check it out):

https://github.com/ccorcos/graph-explorer

**Comment by sverweij** [comment]
@ccorcos I've checked out graph-explorer. That's definitely an interesting way to explore a graph - and worth considering as a future feature (or as a separate project). I'm a bit curious to see how well the [api](https://github.com/sverweij/dependency-cruiser/blob/develop/doc/api.md) would work to create this. 

Thanks for this.

**Comment by ccorcos** [comment]
Interesting. That looks much like the CLI API -- you could always use `child_process` to run via the CLI. What I think would be more interesting and useful for a Node.js API is if depcruise returned a datastructure that I can reuse for incrementally crawling. In terms of the graph explorer, I can start at a node and go to a depth of 2. Then when the user clicks another node, I can expand the graph to a depth of 2 from that node. I'm not sure how well this works with the current architecture though. I think searching for parents cant be terribly efficient...
